smartconfig/_registry.py

```diff
@@ -0,0 +1,120 @@
+"""
+Internal register used to reference entries.
+
+Attributes:
+    global_configuration: The full configuration loaded from the YAML file along with all default values for entries.
+    used_paths: A set of paths already containing an entry.
+    overwritten_attributes: A set of paths that have been overwritten by a configuration file.
+    mapping_cache: A mapping of dot-delimited paths to their corresponding mappings in the config.
+"""
+
+import copy
+from typing import Any, Dict, Hashable, Mapping, MutableMapping, Set
+
+import smartconfig
+from .exceptions import ConfigurationError, ConfigurationKeyError
+
+global_configuration: Dict[Hashable, Any] = {}
+used_paths: Set["smartconfig.ConfigEntry"] = set()
+overwritten_attributes: Set[str] = set()
+mapping_cache: Dict[str, Mapping] = {}
+
+
+def _get_child_node(node_name: str, root: Any):
+    """
+    Retrieve the value of the node `node_name` which is a child of the `root` node.
+
+    Args:
+        node_name: The name of the child node to retrieve.
+        root: The parent node of the node to retrieve.
+
+    Returns:
+        The value of the child node.
+
+    Raises:
+        ConfigurationError: The parent is not a mapping node.
+        ConfigurationKeyError: The child node does not exist.
+    """
+    if not isinstance(root, Mapping):
+        raise ConfigurationError(f"Cannot retrieve node {node_name!r}: its parent is not a mapping node.")
+
+    if node_name not in root:
+        raise ConfigurationKeyError(f"Cannot retrieve node {node_name!r}: it does not exist.")
+
+    node = root[node_name]
+
+    # Use __get__ if the node is a descriptor.
+    if hasattr(node, '__get__'):
+        node = node.__get__()
+
+    return node
+
+
+def unload_defaults(path: str) -> None:
+    """
+    Remove the default values of all children of `path`.
+
+    Args:
+        path: The dot-delimited path to the parent node.
+    """
+    try:
+        node = get_node(path)
+    except (ConfigurationError, ConfigurationKeyError):
+        return
+
+    if not isinstance(node, MutableMapping):
+        return
+
+    for attribute in copy.copy(node):
+        if attribute not in overwritten_attributes:
+            node.pop(attribute)
```

Comment:
Should we just leave the mapping empty or should it be removed if all its children are removed? Removing it may lead to a better error. An attribute missing is a consequence of the entry being unloaded, so an error stating the entry is missing seems ideal. I believe this only matters because we have the reference descriptor.

Reply:
In a similar vein, should the node be removed from `mapping_cache`? If not, there could be an optimisation opportunity by storing a sentinel value indicating the entry has been deleted. This would be quicker than having to traverse the nested dictionaries only to find the node doesn't exist.

Reply:
> Should we just leave the mapping empty or should it be removed if all its children are removed? Removing it may lead to a better error. An attribute missing is a consequence of the entry being unloaded, so an error stating the entry is missing seems ideal. I believe this only matters because we have the reference descriptor.

We could delete the node if it is empty, sure. But if we ever overwrote an attribute, we can't do that.



> In a similar vein, should the node be removed from `mapping_cache`? If not, there could be an optimisation opportunity by storing a sentinel value indicating the entry has been deleted. This would be quicker than having to traverse the nested dictionaries only to find the node doesn't exist.

This sounds a bit like premature optimisation. We don't rely on exceptions in our code except in case of actual failure, if they aren't optimised I don't think it is an issue. 

Thoughts?

Reply:
> We could delete the node if it is empty, sure. But if we ever overwrote an attribute, we can't do that.

Yeah, good point. I think it'd be a bad idea to have inconsistent behaviour. What happens currently when there is a reference to an attribute with a configured value, but the corresponding class is not loaded?

Reply:
> This sounds a bit like premature optimisation. We don't rely on exceptions in our code except in case of actual failure, if they aren't optimised I don't think it is an issue.

You may be right, but I'll have to look into the flow of execution more since I don't remember now. In any case, we can leave this optimisation for another time.

Reply:
I haven't tested it yet but it should still resolve it since it is in the global configuration.

Reply:
I'm not sure if that should be allowed, but I'm fine with leaving it as-is for now.

Reply:
The cache is still something that needs to be looked into in the future. It seems safe to not delete dicts from the cache since `_update_mapping` is careful to actually update existing dicts rather than overwriting them. It's important that the cache does not end up out of sync with what's in the global config.

That being said, keeping values in cache means more memory is used. Keeping or deleting the cache is a matter of a time vs. space trade-off. Deleting manually wouldn't be straight forward since the paths stored are flattened. Weak references could be stored, but this isn't exactly trivial to implement since creating weak references of `dict` (i.e. storing dicts in a `WeakValueDictionary`) is unsupported unless `dict` is subclassed.

> In a similar vein, should the node be removed from mapping_cache? If not, there could be an optimisation opportunity by storing a sentinel value indicating the entry has been deleted. This would be quicker than having to traverse the nested dictionaries only to find the node doesn't exist.

I still don't think this optimisation is completely unfounded, but it's likely too rare of a case to justify implementing it. The proposed implementation has several flaws anyway, which I won't go into.